src/ray/raylet/scheduling/policy/node_label_scheduling_policy.cc

```diff
@@ -0,0 +1,195 @@
+  int node_index = distribution(gen_);
+  auto it = candidate_nodes.begin();
+  std::advance(it, node_index);
+  return it->first;
+}
+
+void NodeLabelSchedulingPolicy::FilterNodesByHardExpressions(
+    absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const rpc::LabelMatchExpressions &expressions) const {
+  for (const auto &expression : expressions.expressions()) {
+    if (candidate_nodes.empty()) {
+      return;
+    }
+
+    for (auto iter = candidate_nodes.begin(); iter != candidate_nodes.end(); iter++) {
+      if (!IsNodeMatchLabelExpression(*(iter->second), expression)) {
+        candidate_nodes.erase(iter);
+      }
+    }
+  }
+}
+
+absl::flat_hash_map<scheduling::NodeID, const Node *>
+NodeLabelSchedulingPolicy::FilterNodesBySoftExpressions(
+    const absl::flat_hash_map<scheduling::NodeID, const Node *> &candidate_nodes,
+    const rpc::LabelMatchExpressions &expressions) const {
+  auto match_all_soft_nodes = candidate_nodes;
+  for (auto it = match_all_soft_nodes.begin(); it != match_all_soft_nodes.end(); it++) {
+    for (const auto &expression : expressions.expressions()) {
+      if (!IsNodeMatchLabelExpression(*(it->second), expression)) {
+        match_all_soft_nodes.erase(it);
+        break;
+      }
+    }
+  }
+
+  return match_all_soft_nodes;
+}
+
+bool NodeLabelSchedulingPolicy::IsNodeMatchLabelExpression(
+    const Node &node, const rpc::LabelMatchExpression &expression) const {
+  const auto &key = expression.key();
+  const auto &match_operator = expression.operator_();
+  if (match_operator.has_label_exists()) {
+    if (IsNodeLabelKeyExists(node, key)) {
+      return true;
+    }
+  } else if (match_operator.has_label_does_not_exist()) {
+    if (!IsNodeLabelKeyExists(node, key)) {
+      return true;
+    }
+  } else if (match_operator.has_label_in()) {
+    std::vector<std::string> values;
+    for (const auto &value : match_operator.label_in().values()) {
+      values.push_back(value);
+    }
+    if (IsNodeLabelInValues(node, key, values)) {
+      return true;
+    }
+  } else if (match_operator.has_label_not_in()) {
+    std::vector<std::string> values;
+    for (const auto &value : match_operator.label_not_in().values()) {
+      values.push_back(value);
+    }
+    if (!IsNodeLabelInValues(node, key, values)) {
+      return true;
+    }
+  } else {
+    RAY_CHECK(false) << "Node label match operator type must be one of `label_in`、"
+                        "`label_not_in`、`label_exists` or `label_does_not_exist`.";
+  }
+  return false;
+}
+
+bool NodeLabelSchedulingPolicy::IsNodeLabelKeyExists(const Node &node,
+                                                     const std::string &key) const {
+  return node.GetLocalView().labels.contains(key);
+}
+
+bool NodeLabelSchedulingPolicy::IsNodeLabelInValues(
+    const Node &node,
+    const std::string &key,
+    const std::vector<std::string> &values) const {
```

Comment:
This can be std::set so we can just do `values.contains(node_labels.at(key))`

Reply:
Fixed